dali/core/mm/pinned_pool_test.cc

```diff
@@ -0,0 +1,154 @@
+    // not just some streams.
+    void *mem2 = pool.allocate_async(N, sv2);
+    auto e = cudaStreamQuery(s1);
+    EXPECT_NE(e, cudaErrorNotReady) << "Synchronization should have occurred";
+    if (e != cudaErrorNotReady) {
+      CUDA_CALL(cudaGetLastError());
+    }
+    EXPECT_EQ(mem1, mem2) << "Memory should have been stolen from the stream1 after it's finished";
+    pool.deallocate_async(mem2, N, sv2);
+  }
+  upstream.check_leaks();
+}
+
+TEST(MMPinnedAlloc, SyncCrossDevice) {
+  test_pinned_resource upstream;
+  int ndev = 0;
+  CUDA_CALL(cudaGetDeviceCount(&ndev));
+  if (ndev < 2) {
+    GTEST_SKIP() << "This test requires at least 2 CUDA devices.";
+  } else {
+    CUDAStream s1, s2;
+    DeviceGuard dg(0);
+    s1 = CUDAStream::Create(true);
+    cudaSetDevice(1);
+    s2 = CUDAStream::Create(true);
+    cudaSetDevice(0);
+    stream_view sv1(s1), sv2(s2);
+    const int N = 1<<24;
+    async_pool_base<memory_kind::pinned> pool(&upstream, true);
+    void *mem1 = pool.allocate_async(N, sv1);
+    CUDA_CALL(cudaMemsetAsync(mem1, 0, N, s1));
+    pool.deallocate_async(mem1, N, sv1);
+    // We've requested a large chunk (16MiB) of memory - that memory is not going
+    // to be readily available, but the pool is configured with "avoid upstream" option
+    // and therefore will wait for the pending deallocations to complete - this is still
+    // lighter than calling cudaMallocHost, which would implicitly synchronize all devices,
+    // not just some streams.
+    void *mem2 = pool.allocate_async(N, sv2);
+    auto e = cudaStreamQuery(s1);
+    EXPECT_NE(e, cudaErrorNotReady) << "Synchronization should have occurred";
+    if (e != cudaErrorNotReady) {
+      CUDA_CALL(cudaGetLastError());
+    }
+    EXPECT_EQ(mem1, mem2) << "Memory should have been stolen from the stream1 after it's finished";
+    pool.deallocate_async(mem2, N, sv2);
+  }
+  upstream.check_leaks();
+}
+
+TEST(MMPinnedAlloc, FreeOnAnotherDevice) {
+  test_pinned_resource upstream;
+  int ndev = 0;
+  CUDA_CALL(cudaGetDeviceCount(&ndev));
+  if (ndev < 2) {
+    GTEST_SKIP() << "This test requires at least 2 CUDA devices.";
+  } else {
+    CUDAStream s1, s2;
+    DeviceGuard dg(0);
+    s1 = CUDAStream::Create(true);
+    cudaSetDevice(1);
+    s2 = CUDAStream::Create(true);
+    cudaSetDevice(0);
+    stream_view sv1(s1), sv2(s2);
+    const int N = 1<<24;
+    async_pool_base<memory_kind::pinned> pool(&upstream, true);
+    void *mem1 = pool.allocate_async(N, sv1);
+    CUDA_CALL(cudaMemsetAsync(mem1, 0, N, s1));
+    cudaStreamSynchronize(s1);
+    // don't set device - it should be inferred from the stream
+    pool.deallocate_async(mem1, N, sv2);
+    // now set the device and allocate
+    cudaSetDevice(1);
+    void *mem2 = pool.allocate_async(N, sv2);
+    EXPECT_EQ(mem1, mem2) << "Memory should have been moved to stream2 on another device.";
+    pool.deallocate_async(mem2, N, sv2);
```

Comment:
```suggestion
    {
      DeviceGuard dg2(1);
      void *mem2 = pool.allocate_async(N, sv2);
      EXPECT_EQ(mem1, mem2) << "Memory should have been moved to stream2 on another device.";
      pool.deallocate_async(mem2, N, sv2);
    }
```

Reply:
I don't think it matters - the only reason I use device guard at all is to restore the default device at the end of the test, even if it fails.